Plutarch/FFI.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.FFI (
+  foreignExport,
+  foreignImport,
+  unsafeForeignExport,
+  unsafeForeignImport,
+) where
+
+import Data.Kind (Constraint, Type)
+import Data.Text (Text)
+import GHC.TypeLits (TypeError)
+import qualified GHC.TypeLits as TypeLits
+import qualified Generics.SOP as SOP
+import Plutarch.Bool (PBool)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  ClosedTerm,
+  PDelayed,
+  PType,
+  RawTerm (RCompiled),
+  Term (Term),
+  TermResult (TermResult),
+  asClosedRawTerm,
+  compile',
+  (:-->),
+ )
+import Plutarch.Internal.PlutusType (PlutusType (PInner))
+import Plutarch.String (PString)
+import Plutus.V1.Ledger.Scripts (Script (unScript), fromCompiledCode)
+import PlutusTx.Code (CompiledCode, CompiledCodeIn (DeserializedCode))
+import PlutusTx.Prelude (BuiltinString)
+import UntypedPlutusCore (fakeNameDeBruijn)
+import qualified UntypedPlutusCore as UPLC
+
+data ForallPhantom :: Type
+data PhorallPhantom :: PType
+
+foreignExport :: PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom => ClosedTerm p -> CompiledCode t
```

Comment:
I'm honestly not sure why this type is this complex. Why not just do `ClosedTerm p -> ClosedTerm (PlutarchToPlutusTx p)`? You could make type family injective to make it reversible.

Reply:
`PlutarchToPlutusTx PScriptContext` would need to reduce to `ScriptContext`, and so on for all ADTs including user-defined ones.

Reply:
You can not export `ClosedTerm PScriptContext` as `CompiledCode ScriptContext`.

Reply:
That's true, because `PScriptContext` is currently defined via `PDataRecord`. I think we should have two variants of `PScriptContext` and the rest of API: one Data-encoded and the other decoded, i.e. plain.